src/Symfony/Component/Mailer/Transport.php

```diff
@@ -0,0 +1,180 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Mailer;
+
+use Psr\Log\LoggerInterface;
+use Symfony\Component\EventDispatcher\EventDispatcherInterface;
+use Symfony\Component\Mailer\Bridge\Amazon;
+use Symfony\Component\Mailer\Bridge\Google;
+use Symfony\Component\Mailer\Bridge\Mailchimp;
+use Symfony\Component\Mailer\Bridge\Mailgun;
+use Symfony\Component\Mailer\Bridge\Postmark;
+use Symfony\Component\Mailer\Bridge\Sendgrid;
+use Symfony\Component\Mailer\Exception\InvalidArgumentException;
+use Symfony\Component\Mailer\Exception\LogicException;
+use Symfony\Component\Mailer\Transport\TransportInterface;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+
+/**
+ * @author Fabien Potencier <fabien@symfony.com>
+ *
+ * @experimental in 4.3
+ */
+class Transport
+{
+    public static function fromDsn(string $dsn, EventDispatcherInterface $dispatcher = null, HttpClientInterface $client = null, LoggerInterface $logger = null): TransportInterface
+    {
+        // failover?
+        $dsns = preg_split('/\s++\|\|\s++/', $dsn);
+        if (\count($dsns) > 1) {
+            $transports = [];
+            foreach ($dsns as $dsn) {
+                $transports[] = self::createTransport($dsn, $dispatcher, $client, $logger);
+            }
+
+            return new Transport\FailoverTransport($transports);
+        }
+
+        // round robin?
+        $dsns = preg_split('/\s++&&\s++/', $dsn);
+        if (\count($dsns) > 1) {
+            $transports = [];
+            foreach ($dsns as $dsn) {
+                $transports[] = self::createTransport($dsn, $dispatcher, $client, $logger);
+            }
+
+            return new Transport\RoundRobinTransport($transports);
+        }
+
+        return self::createTransport($dsn, $dispatcher, $client, $logger);
+    }
+
+    private static function createTransport(string $dsn, EventDispatcherInterface $dispatcher = null, HttpClientInterface $client = null, LoggerInterface $logger = null): TransportInterface
+    {
+        if (false === $parsedDsn = parse_url($dsn)) {
+            throw new InvalidArgumentException(sprintf('The "%s" mailer DSN is invalid.', $dsn));
+        }
+
+        if (!isset($parsedDsn['host'])) {
+            throw new InvalidArgumentException(sprintf('The "%s" mailer DSN must contain a mailer name.', $dsn));
+        }
+
+        $user = \urldecode($parsedDsn['user'] ?? '');
+        $pass = \urldecode($parsedDsn['pass'] ?? '');
+        \parse_str($parsedDsn['query'] ?? '', $query);
+
+        switch ($parsedDsn['host']) {
```

Comment:
I didn't look at other places, but I think it is too bad that this is a "fixed" list, there are still problems such as https://github.com/symfony/swiftmailer-bundle/issues/217. So if your mail provider / transport is not implemented here, you can't have a dsn-like custom transport.

I could be wrong though, and that'd be great. :}

Reply:
wow, lots of negative statements here :( what would be great? you didn't explain your idea? on my side, I don't know how this list could be otherwise than fixed.

Reply:
Not a negative comment, it's just something I would like to have, but if it's not there, I'm still going to use it. :}

I would have liked something like an application of the decision pattern, e.g have a bunch of transports (as we have now), which each tells if they support a host / a scheme (a dsn actually), so I can register *any* transport and use it through its DSN.

E.g for the mailjet thing I linked, have the possibility to use `mailjet://...` or `api://mailjet....` at least. Currently (if I understood it right), you can't and are bound to use one of the supported transports...

Once again, I may be wrong. Anyway, even without that, the component is awesome. :}

Reply:
You mean like a factory strategy? I agree, but this is more about usage without the Framework I guess done a helper class rather than something that is fully concrete.

Maybe the static method can be kept for the standalone but allow the registering of other transporters. Like https://github.com/rollerworks/search/blob/master/lib/Core/Loader/InputProcessorLoader.php#L46

Reply:
Yep, sorry I meant the factory strategy indeed.